internal/host/host_test.go

```diff
@@ -192,17 +192,73 @@ var _ = Describe("update_role", func() {
 		})
 	})
 
-	It("update role master to worker", func() {
-		By("update role worker to master")
-		host = hostutil.GenerateTestHost(id, infraEnvID, clusterID, models.HostStatusKnown)
-		Expect(db.Create(&host).Error).ShouldNot(HaveOccurred())
-		Expect(state.UpdateRole(ctx, &host, models.HostRoleMaster, nil)).NotTo(HaveOccurred())
-		h := hostutil.GetHostFromDB(id, infraEnvID, db)
-		Expect(h.Role).To(Equal(models.HostRoleMaster))
-		By("update role master to worker")
-		Expect(state.UpdateRole(ctx, &h.Host, models.HostRoleWorker, nil)).NotTo(HaveOccurred())
-		h = hostutil.GetHostFromDB(id, infraEnvID, db)
-		Expect(h.Role).To(Equal(models.HostRoleWorker))
+	Context("by requested role", func() {
+		var h *common.Host
+
+		Context("from master in day1 flow", func() {
+			BeforeEach(func() {
+				host = hostutil.GenerateTestHost(id, infraEnvID, clusterID, models.HostStatusKnown)
+				Expect(db.Create(&host).Error).ShouldNot(HaveOccurred())
+				Expect(state.UpdateRole(ctx, &host, models.HostRoleMaster, nil)).NotTo(HaveOccurred())
+				h = hostutil.GetHostFromDB(id, infraEnvID, db)
+				Expect(h.Role).To(Equal(models.HostRoleMaster))
+			})
+
+			It("to master successfuly", func() {
+				Expect(state.UpdateRole(ctx, &h.Host, models.HostRoleMaster, nil)).NotTo(HaveOccurred())
+				h = hostutil.GetHostFromDB(id, infraEnvID, db)
+				Expect(h.Role).To(Equal(models.HostRoleMaster))
+			})
+			It("to worker successfuly", func() {
+				Expect(state.UpdateRole(ctx, &h.Host, models.HostRoleWorker, nil)).NotTo(HaveOccurred())
+				h = hostutil.GetHostFromDB(id, infraEnvID, db)
+				Expect(h.Role).To(Equal(models.HostRoleWorker))
+			})
+			It("to bootstrap successfuly", func() {
+				Expect(state.UpdateRole(ctx, &h.Host, models.HostRoleBootstrap, nil)).NotTo(HaveOccurred())
+				h = hostutil.GetHostFromDB(id, infraEnvID, db)
+				Expect(h.Role).To(Equal(models.HostRoleBootstrap))
+			})
+			It("to chocobomb with failure", func() {
```

Comment:
this test is not future-proof, what if someone adds a chocobomb role in the future? this test will break!

Reply:
Your test should instead iterate all possible roles using reflection, then dynamically generate a role that necessarily doesn't exist, this way you can save this theoretical future chocobomb dev a few seconds of fixing unit tests

Reply:
(i sure hope you don't think I was serious in any of my messages)

But in this case the unit-test is breaking for irrelevant reasons. There's nothing wrong with adding a chocobomb role.Yet the unit-test failed. The developer has become aware of the test, but should they have been aware of it?

Reply:
Let me step in the shoes of a theoretical dev, Joshua, from the future who tries to add a chocobomb role. We will walk through a proof that existence of this test is only of a benefit...

Joshua decides that chocobomb is a valid role for all the flows, so he goes to the function IsRoleValid and does his stuff. At this moment if Joshua is not experienced enough, he just commits the code and prays he gets LTGM without anyone noticing he missed the unit test.

If Joshua is an experienced developer, he will immediately start writing a unit test to cover the fact that chocobomb is now allowed role. At this moment he notices that I created a test for the same role and just piggybacks on it by modifying the condition. In this case my test helped Joshua because he doesn't need to write a new case from scratch, he can just edit the test that I have created.

So by the reasoning above I have proven that existence of the chocobomb test can be only a benefit. For inexperienced Joshua it will be a learning experience as his build will fail and force him to at least look at unit test (and hopefully remember for the future that those are a must). For experienced Joshua it will save him something between 1 up to 2 lines of code. By some quick calculation this is a save of around 3 kg of CO2 emission but I leave the calculation to the reader as it's not relevant to the proof.

To sum up, we have proven that existence of a test for chocobomb role can be only of a help.